### DISCUSSION TITLE
string/string-source

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#213] [Announcements]
# string/string-source

本文将对Java String类的源码进行深入分析，帮助你理解字符串在Java中的实现原理、操作方式和性能优化策略。通过详细解读String源码，我们将揭示字符串类的内部机制，掌握如何高效地进行字符串操作，以及在实际编程中如何应用这些知识来提高代码质量和性能。

https://tobebetterjavaer.com/string/string-source.html

**Comment by KayunZhang** [comment]
int max = sourceOffset + (sourceCount - targetCount);

**Reply by KayunZhang** [reply]
这个代码是不是有问题啊  虽然它是个内部方法  不被调用  但如果真传非0的正值进去的话 就算能找到下标  也是错的吧 需要加上sourceOffset，如果没有找到，还会抛错。是我没理解开发者的意图么。。。

**Reply by KayunZhang** [reply]
哦  对了  这是indexOf源码内的

**Reply by Tim1Wu** [reply]
这里是有作者的思考的，首先这是一个静态方法。可以使用类名直接调用静态方法。其次sourceOffset + (sourceCount - targetCount);这段代码的含义无非是算从sourceOffset开始需要搜索的最大偏移量。就算是乱填，该方法也不会抛错，无非返回一个-1表示没查到。

**Comment by yclwk** [comment]
String str = "Hello, world!";
String prefix = str.substring(0, 5);  // 提取前5个字符，即 "Hello,"
应该为Hello吧？

**Reply by Eivae** [reply]
应该是的

**Comment by absoultly-classical** [comment]
为什么indexof()方法不用kmp算法？

**Comment by LiSeries** [comment]
String str = "Hello, world!";
String subStr = str.substring(7, 12);  // 从第7个字符（包括）提取到第12个字符（不包括）
System.out.println(subStr);  // 输出 "world"
应该是从索引7，也就是第八个字符开始提取，否则输出的就是" world"。后面几个例子也是同样的问题

**Comment by Akinator365** [comment]
其中，s[i]表示字符串 s 中第 i 个字符的 ASCII 码值，^表示幂运算。这里如果是中文字符的话，应该是Unicode编码吧
